File: 2023/AoC23-15ab.swift
postfix operator +

extension Int {
  static postfix func + (n: Int) -> Int {
    return n + 1
  }
}

extension Array {
  mutating func retain(_ cond: (Element) -> Bool) {
    self = self.filter(cond)
  }
}

struct Lens {
  let label: String
  var focalLength: Int

  init(_ label: String, _ focalLength: Int) {
    self.label = label
    self.focalLength = focalLength
  }
}

func hashAlg(_ s: String) -> Int {
  return s.reduce(0) { (($0 + Int($1.asciiValue!)) * 17) % 256 }
}

func part2(_ lines: [String]) -> Int {
  var boxes: [[Lens]] = [[Lens]](repeating: [], count: 256)

  for line in lines {
    if let i = line.firstIndex(of: "=") {
      let label = String(line[..<i])
      let focalLength = Int(line[line.index(after: i)...])!

      let pos = hashAlg(label)
      switch boxes[pos].firstIndex(where: { $0.label == label } ) {
        case let .some(j):
          boxes[pos][j].focalLength = focalLength
        default:
          boxes[pos].append(Lens(label, focalLength))
      }
    } else {
      let label = String(line.dropLast())
      boxes[hashAlg(label)].retain { $0.label != label }
    }
  }

  return boxes.enumerated().reduce(0) {
    let (i, box) = $1
    return $0 + box.enumerated().reduce(0) {
      let (j, lens) = $1
      return $0 + i+ * j+ * lens.focalLength
    }
  }
}

let lines = readLine()!.split(separator: ",").map(String.init)

print("Part 1 answer:", lines.reduce(0) { $0 + hashAlg($1) })

print("Part 2 answer:", part2(lines))

File: 2023/AoC23-15ab.cpp
#include <iostream>
#include <string>
#include <vector>
#include <fstream>

class Lens {
  public:
    std::string label;
    int focal_length;

  Lens(std::string label, int focal_length) {
    this->label = label;
    this->focal_length = focal_length;
  }
};

std::vector<std::string> get_lines(std::string input) {
  std::vector<std::string> lines = {};

  if (input[input.size()-1] != ',') input.push_back(',');
  size_t pos = 0;
  while ((pos = input.find(',')) != std::string::npos) {
    lines.push_back(input.substr(0, pos));
    input.erase(0, pos + 1);
  }
  
  return lines;
}

int hash(std::string s) {
  int out = 0;
  for (char c : s) {
    out = ((out + c) * 17) % 256;
  }
  
  return out;
}

int main() {
  std::string temp;
  std::ifstream input("input.txt");
  input >> temp;
  input.close();

  const std::vector<std::string> lines = get_lines(temp);
  
  int total1 = 0;
  for (std::string s : lines) {
    total1 += hash(s);
  }
  std::cout << "Part 1 answer: " << total1 << "\n";

  std::vector<std::vector<Lens>> boxes(256, std::vector<Lens>{});
  
  for (std::string line : lines) {
    size_t pos = 0;
    if ((pos = line.find('=')) != std::string::npos) {
      const std::string label = line.substr(0, pos);
      const int box = hash(label);
      const int focal_length = std::stoi(line.substr(pos + 1, 1));
      
      bool found = false;
      for (int i = 0; i < boxes[box].size(); i++) {
        if (boxes[box][i].label == label) {
          boxes[box][i].focal_length = focal_length;
          found = true;
          break;
        }
      }
      
      if (!found) {
        boxes[box].push_back(Lens(label, focal_length));
      }
    } else {
      const std::string label = line.substr(0, line.size()-1);
      const int box = hash(label);
      
      for (int i = 0; i < boxes[box].size(); i++) {
        if (boxes[box][i].label == label) {
          boxes[box].erase(boxes[box].begin() + i);
          break;
        }
      }
    }
  }

  int total2 = 0;
  for (int i = 0; i < boxes.size(); i++) {
    for (int j = 0; j < boxes[i].size(); j++) {
      total2 += (i + 1) * (j + 1) * boxes[i][j].focal_length;
    }
  }
  std::cout << "Part 2 answer: " << total2 << "\n";
}
